runtime/marshaler_registry.go

```diff
@@ -5,132 +5,87 @@ import (
 	"net/http"
 )
 
-const mimeWildcard = "*"
+// MIMEWildCard is the fallback MIME type used for requests which do not match
+// a registered MIME type.
+const MIMEWildcard = "*"
 
 var (
-	defaultMarshaler = &JSONPb{OrigName: true}
-
+	acceptHeader      = http.CanonicalHeaderKey("Accept")
 	contentTypeHeader = http.CanonicalHeaderKey("Content-Type")
+
+	defaultMarshaler = &JSONPb{OrigName: true}
 )
 
 // MarshalerForRequest returns the inbound/outbound marshalers for this request.
 // It checks the registry on the ServeMux for the MIME type set by the Content-Type header.
 // If it isn't set (or the request Content-Type is empty), checks for "*".
+// If that isn't set, uses the ServerMux's InboundMarshaler/OutboundMarshaler.
 // If there are multiple Content-Type headers set, choose the first one that it can
 // exactly match in the registry.
 // Otherwise, it follows the above logic for "*"/InboundMarshaler/OutboundMarshaler.
 func MarshalerForRequest(mux *ServeMux, r *http.Request) (inbound Marshaler, outbound Marshaler) {
-	headerVals := append(append([]string(nil), r.Header[contentTypeHeader]...), "*")
-
-	for _, val := range headerVals {
-		m := mux.marshalers.lookup(val)
-		if m != nil {
-			if inbound == nil {
-				inbound = m.inbound
-			}
-			if outbound == nil {
-				outbound = m.outbound
-			}
+	for _, acceptVal := range r.Header[acceptHeader] {
+		if m, ok := mux.marshalers.mimeMap[acceptVal]; ok {
+			outbound = m
+			break
 		}
-		if inbound != nil && outbound != nil {
-			// Got them both, return
-			return inbound, outbound
+	}
+
+	for _, contentTypeVal := range r.Header[contentTypeHeader] {
+		if m, ok := mux.marshalers.mimeMap[contentTypeVal]; ok {
+			outbound = m
+			break
 		}
 	}
+
 	if inbound == nil {
-		inbound = defaultMarshaler
+		inbound = mux.marshalers.mimeMap[MIMEWildcard]
 	}
 	if outbound == nil {
-		outbound = defaultMarshaler
+		outbound = mux.marshalers.mimeMap[MIMEWildcard]
```

Comment:
Could you try falling back to `mux.marshalers.mimeMap[contentTypeHeader]` before `MIMEWildcard`?